### ISSUE TITLE
[Feature]: Tree attention about Speculative Decoding

### ISSUE THREAD

**Issue by yukavio** [vllm-project/vllm#3960]
### 🚀 The feature, motivation and pitch

I want to implement tree attention for vllm mentioned in [RoadMap](https://github.com/vllm-project/vllm/issues/3861). But I don’t know whether I should implement it based on paged-attention kernel implemented in vllm or FlashInfer due to I found we plan to replace this kernel in this [PR.](https://github.com/vllm-project/vllm/pull/2772)

**Comment by cadedaniel** [comment]
Thanks for your interest in contributing! FYI tree attention is a bit complicated to implement with non-contiguous KV cache, since intra-block attention masking has not been implemented anywhere AFAIK. We can get around this by limiting vLLM to block size of 1, but this makes it difficult to optimize latency of verification as we limit the allowed vLLM configuration space.

The way I'd recommend going about this is to implement intra-block attention masking first, then integrate it with vLLM. This is the surefire way to obtain the best latency reduction possible in vLLM. The steps as follows:

* (Kernel) Support attention mask inside single block (M)
* (Worker) Support attention mask in Worker API (S-M)
* (Spec Decode Framework) Propose, score, and verify top-k candidates (M) (e.g. implement replacement for [this](https://github.com/vllm-project/vllm/blob/b3104b2a10ab7cb7532442177ae0d0c40acf9d03/vllm/spec_decode/batch_expansion.py#L19))
* (Spec Decode Framework) Defragment accepted KV blocks (S-M)

After the remaining open sourcing work is complete, I'll add some documentation for this.

**Comment by reyoung** [comment]
@cadedaniel @yukavio 

Tree attention mechanisms can also be utilized to generate multiple outcomes from the same prompt by varying the seeds.

This approach is an effective strategy to ensure the stability of results produced by Large Language Models (LLMs). For instance, when employing an LLM as a scoring tool to derive metrics, one could sample the LLM's outputs multiple times. By averaging these samples, a more reliable result can be obtained.

---

This feature might become available following the implementation of tree attention mechanisms.

**Comment by yukavio** [comment]
@cadedaniel 
    Thanks for your reply. I have read your document and it seems that the key to the problem is that each token in the prefill phase requires a loop and calculation of the entire kv-cache. 
    I think this problem can be solved by storing all prefill tokens for a certain seq under the same adjacent address, instead of treating them as different seqs after expansion. In this way, we can perform calculations efficiently through tensor-core with a specific attention mask. 
    But in this way, we should organize the prefill token in one sequence instead of multiple of sequence.

**Comment by cadedaniel** [comment]
@yukavio you should talk with @LiuXiaoxuanPKU , who is adding MQA scoring to vLLM

**Comment by jon-chuang** [comment]
@cadedaniel what is the status of this?

**Comment by jon-chuang** [comment]
I guess it is blocked on this? https://github.com/vllm-project/vllm/pull/5691

**Comment by jon-chuang** [comment]
Anw, for reference, this is what people refer to by tree attention, right? https://arxiv.org/pdf/2305.09781

**Comment by cadedaniel** [comment]
Yep, it is blocked on #5691. https://arxiv.org/pdf/2305.09781 was the first paper discussing this approach to scoring, although [Medusa](https://sites.google.com/view/medusa-llm) did a better job in popularizing it.

It is not a scheduling nightmare; the scheduler today with lookahead scheduling already supports tree attention. The complexity is mapping the linear lookahead blocks to the tree structure such that 1) the attention operation has correct causal masks and 2) the linear lookahead blocks are defragmented after the accepted tokens are determined.

See https://docs.google.com/document/d/1Z9TvqzzBPnh5WHcRwjvK2UEeFeq5zMZb5mFE8jR0HCs/edit#heading=h.1fjfb0donq5a for more information on lookahead scheduling.

**Comment by jon-chuang** [comment]
So it is MQA with a per-branch mask while speculating?

So at the kernel level one should compact the non-accepted tokens away after verification, right? And that would be the thrust of #5691?

**Comment by cadedaniel** [comment]
Yep, although you can do compaction in PyTorch without custom kernel

**Comment by Siegfried-qgf** [comment]
> @cadedaniel 谢谢你的回复。我看了你的文档，似乎问题的关键在于score阶段每个token都需要循环计算整个kv-cache。 我觉得这个问题可以通过将某个seq的所有pre-score token都存放在同一个相邻地址下来解决，而不是在展开后当成不同的seq。这样我们就可以通过tensor-core配合特定的attention mask高效地进行计算。 但是这样的话，我们应该将pre-score token组织在一个序列中（img中的左边），而不是多个序列中（img中的右边）。 如果你觉得这种组织pre-score token的方式合适，我可以用treetention mask来实现tensor-core cuda kernel。 <img alt="图像" width="584" src="https://private-user-images.githubusercontent.com/67678385/321209566-7db14a76-113e-46fe-8157-3afff4c4e257.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU1OTQyNTcsIm5iZiI6MTcyNTU5Mzk1NywicGF0aCI6Ii82NzY3ODM4NS8zMjEyMDk1NjYtN2RiMTRhNzYtMTEzZS00NmZlLTgxNTctM2FmZmY0YzRlMjU3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTA2VDAzMzkxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJmNGQzYmIyNTE2MmRjMTE5NDRjNTkxNjk0YzA0ZjdhY2RiYzE0YzdkM2JhNTI4YzBjOGU1MWY2ZmIwYWQ3ZTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.4PC23FKzD7kxRmF_QppskFQq-t-r6p5obYEv3AZd4d8">



> @cadedaniel Thanks for your reply. I have read your document and it seems that the key to the problem is that each token in the score phase requires a loop and calculation of the entire kv-cache. I think this problem can be solved by storing all pre-score tokens for a certain seq under the same adjacent address, instead of treating them as different seqs after expansion. In this way, we can perform calculations efficiently through tensor-core with a specific attention mask. But in this way, we should organize the pre-score token in one sequence (left in img) instead of multiple sequences (right in img). <img alt="image" width="584" src="https://private-user-images.githubusercontent.com/67678385/321209566-7db14a76-113e-46fe-8157-3afff4c4e257.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU1OTQ0MzMsIm5iZiI6MTcyNTU5NDEzMywicGF0aCI6Ii82NzY3ODM4NS8zMjEyMDk1NjYtN2RiMTRhNzYtMTEzZS00NmZlLTgxNTctM2FmZmY0YzRlMjU3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTA2VDAzNDIxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTllZDc4ZDQ1YmQ3NzY1NGUyY2FjNzA5YmMxYzYyNDIzMjNiZDAyNjViOTk3NzFkOTllODRmOWY3OWY3MGJlYTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.enQe1bmce4mCW9EyAR_zbZp6xoy7NIMDaTU0mLhOdKM"> If you think this way of organizing pre-score tokens is appropriate, I can implement the tensor-core cuda kernel with tree attention mask.

Nice idea. How is your progress? Are there any difficulties in the implementation process?

**Comment by Siegfried-qgf** [comment]
> @cadedaniel Thanks for your reply. I have read your document and it seems that the key to the problem is that each token in the score phase requires a loop and calculation of the entire kv-cache. I think this problem can be solved by storing all pre-score tokens for a certain seq under the same adjacent address, instead of treating them as different seqs after expansion. In this way, we can perform calculations efficiently through tensor-core with a specific attention mask. But in this way, we should organize the pre-score token in one sequence (left in img) instead of multiple sequences (right in img). <img alt="image" width="584" src="https://private-user-images.githubusercontent.com/67678385/321209566-7db14a76-113e-46fe-8157-3afff4c4e257.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU1OTQ0MzMsIm5iZiI6MTcyNTU5NDEzMywicGF0aCI6Ii82NzY3ODM4NS8zMjEyMDk1NjYtN2RiMTRhNzYtMTEzZS00NmZlLTgxNTctM2FmZmY0YzRlMjU3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTA2VDAzNDIxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTllZDc4ZDQ1YmQ3NzY1NGUyY2FjNzA5YmMxYzYyNDIzMjNiZDAyNjViOTk3NzFkOTllODRmOWY3OWY3MGJlYTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.enQe1bmce4mCW9EyAR_zbZp6xoy7NIMDaTU0mLhOdKM"> If you think this way of organizing pre-score tokens is appropriate, I can implement the tensor-core cuda kernel with tree attention mask.

Nice idea. How is your progress? Are there any difficulties in the implementation process?
